File: by-pg-flutter/lib/services/db_service.dart
import 'package:postgres/postgres.dart';

/// PostgreSQL-Service für test2view und buchpos (wie C# by-pg-grid).
class DbService {
  static const String host = '192.168.207.160';
  static const int port = 5432;
  static const String database = 'RK2';
  static const String username = 'postgres';
  static const String password = 'Ole1brumm';

  static Endpoint get _endpoint => Endpoint(
        host: host,
        port: port,
        database: database,
        username: username,
        password: password,
      );

  static ConnectionSettings get _settings => ConnectionSettings(
        sslMode: SslMode.disable,
      );

  /// Lädt alle Zeilen aus public.test2view.
  static Future<List<Map<String, dynamic>>> loadTest2View() async {
    final conn = await Connection.open(_endpoint, settings: _settings);
    try {
      final result = await conn.execute(
        Sql.named('SELECT * FROM public.test2view'),
      );
      return result.map((row) => row.toColumnMap()).toList();
    } finally {
      await conn.close();
    }
  }

  /// Lädt alle Zeilen aus public.buchpos (für Update).
  static Future<List<Map<String, dynamic>>> loadBuchpos() async {
    final conn = await Connection.open(_endpoint, settings: _settings);
    try {
      final result = await conn.execute(
        Sql.named('SELECT * FROM public.buchpos'),
      );
      return result.map((row) => row.toColumnMap()).toList();
    } finally {
      await conn.close();
    }
  }

  /// UPDATE buchpos für eine Zeile (wie ButtonSavePreview / update row).
  static Future<void> updateBuchposRow({
    required int pk1,
    required String? bvh,
    required String? test,
    required double betrag,
    required String? art,
    required String? gewerk,
  }) async {
    final conn = await Connection.open(_endpoint, settings: _settings);
    try {
      await conn.execute(
        Sql.named(
          'UPDATE public.buchpos SET bvh = @bvh, test = @test, betrag = @betrag, art = @art, gewerk = @gewerk WHERE pk1 = @pk1',
        ),
        parameters: {
          'bvh': bvh,
          'test': test,
          'betrag': betrag,
          'art': art,
          'gewerk': gewerk,
          'pk1': pk1,
        },
      );
    } finally {
      await conn.close();
    }
  }
}

File: by-pg-ruby/lib/db_service.rb
# frozen_string_literal: true

require 'pg'

# PostgreSQL-Service für test2view und buchpos (wie by-pg-grid).
module DbService
  HOST     = '192.168.207.160'
  PORT     = 5432
  DBNAME   = 'RK2'
  USER     = 'postgres'
  PASSWORD = 'Ole1brumm'

  def self.conn
    PG.connect(
      host: HOST,
      port: PORT,
      dbname: DBNAME,
      user: USER,
      password: PASSWORD
    )
  end

  # Lädt alle Zeilen aus public.test2view.
  # Rückgabe: Array von Hashes (Spaltenname => Wert)
  def self.load_test2view
    c = self.conn
    result = c.exec('SELECT * FROM public.test2view')
    cols = result.fields
    rows = result.values.map { |vals| cols.zip(vals).to_h }
    c.close
    rows
  rescue PG::Error => e
    raise "LoadTest2View Fehler: #{e.message}"
  end

  # UPDATE public.buchpos für eine Zeile (wie save row / update row).
  def self.update_buchpos_row(pk1:, bvh:, test:, betrag:, art:, gewerk:)
    c = self.conn
    c.exec_params(
      'UPDATE public.buchpos SET bvh=$1, test=$2, betrag=$3, art=$4, gewerk=$5 WHERE pk1=$6',
      [bvh.to_s, test.to_s, betrag.to_f, art.to_s, gewerk.to_s, pk1.to_i]
    )
    c.close
  rescue PG::Error => e
    raise "UpdateBuchposRow Fehler: #{e.message}"
  end
end
